File: 1608/solution.py
class Solution:
    def specialArray(self, nums: list[int]) -> int:
        nums.sort()
        i = 0
        while i < len(nums):
            nums_left = len(nums) - i
            if nums_left == nums[i]:
                return nums[i]
            if i == 0 and nums_left < nums[i]:
                return nums_left
            if nums_left > nums[i - 1] and nums_left < nums[i]:
                return nums_left
            while i + 1 < len(nums) and nums[i + 1] == nums[i]:
                i += 1
            i += 1
        return -1

File: 1608/solution.java
import java.util.Arrays;

class Solution {
    public int specialArray(int[] nums) {
        Arrays.sort(nums);
        int i = 0;
        while (i < nums.length) {
            int numsLeft = nums.length - i;
            if (numsLeft == nums[i]) {
                return numsLeft;
            }
            int numBefore = i == 0 ? 0 : nums[i - 1];
            if (numsLeft < nums[i] && numsLeft > numBefore) {
                return numsLeft;
            }
            while (i + 1 < nums.length && nums[i + 1] == nums[i]) {
                i++;
            }
            i++;
        }
        return -1;
    }
}

// time O(nlogn)
// space O(1)